keps/sig-apps/4368-support-managed-by-label-for-batch-jobs/README.md

```diff
@@ -0,0 +1,854 @@
+
+The mechanism should be per-Job, because the management cluster might also be
+one of the worker clusters, for two reasons:
+1. Disabling the Job controller per cluster requires access the `kube-controller-manager`
+   manifest. Such access is generally discouraged by cloud providers.
+2. The management cluster may also be a worker. Supporting this scenario is important
+   for smooth transition of Kueue users from a single-cluster to multi-cluster.
+
+Ideally, I prefer the mechanism to be lightweight so that it is easy to be adopted
+by other Job CRDs supported by Kueue
+(see [here](https://github.com/kubernetes-sigs/kueue/blob/6d428f3279a9ca0e204c083dc649dbbc6558db71/config/components/manager/controller_manager_config.yaml#L31-L42)):
+MPIJob, RayJob, JobSet, multiple Kubeflow jobs.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Prior work
+
+This approach of allowing another controller to mirror information between APIs
+is already supported with the `managed-by` label used by
+EndpointSlices (`endpointslice.kubernetes.io/managed-by`) and IPAddresses
+(`ipaddress.kubernetes.io/managed-by`).
+
+#### Graduating directly to Beta
+
+The implementation is simple (see [Design details](#design-details)), short-circuit the
+`syncJob` invocation. Also, since the feature is just label-based, there is
+technically no need to go via Alpha in order to ensure that all
+kube-apiserver instances (in HA setup, see [here](https://kubernetes.io/releases/version-skew-policy/#supported-version-skew))
+recognize the new API, as in case of new fields. A similar approach was used
+to graduate the
+[Pod Index Label for StatefulSets and Indexed Jobs](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/4017-pod-index-label)
+feature with similar characteristics.
+
+Finally, we would like to have the feature soak without unnecessary delay.
+
+### Risks and Mitigations
+
+#### Ecosystem fragmentation due to forks
+
+The mechanism to disable the main Job controller opens the door for users to
+substitute it with a fork. This may create more fragmentation in the community
+as users may prefer to use their forked controllers rather than contribute
+upstream.
+
+First, this risk, to some extent, exists even today as admins with access to the
+control plane can disable by passing `controllers=-job,*` in the manifest for of
+`kube-controller-manager`.
+
+Second, we believe that users who had the need to fork the Job controller
+already introduced dedicated Job CRDs for their needs.
+
+## Design Details
+
+#### API
+
+```golang
+const (
+  ...
+	// LabelManagedBy is used to indicate the controller or entity that manages
+	// an Job.
+	LabelManagedBy = "batch.kubernetes.io/managed-by"
+)
+```
+
+#### Implementation overview
+
+We skip synchronization of the Jobs with the `managed-by` label, if it has any
+different value than `job-controller`.
+
+There is no validation for the values of the field beyond that of standard
+permitted label values.
+
```

Comment:
So the main questions I have for this design:

How is the label going to be added? ie pkg/registry?

What do we do about existing labels? And is there any protections on a user removing or adding a label for this?

I guess that a user probably can only patch the jobs in their namespace so they could probably only disable jobs that they have access too.

Reply:
Last thing, what is the expected behavior if a label is added on an existing job? Will we stop syncing the running job?

Reply:
> How is the label going to be added? ie pkg/registry? default for all jobs?

For MultiKueue, it will be added by the Kueue controller based on its scheduling decision. 

In general, it can be added by end-users or other frameworks.

> default for all jobs?

No defaulting. Discussed it now in alternatives in "Defaulting of the manage-by label for newly created jobs".

> What do we do about existing labels? And is there any protections on a user removing or adding a label for this?

The mutability of labels is controller by `AllowMutableSchedulingDirectives`. I discuss this now in the `Mutability` section.

> I guess that a user probably can only patch the jobs in their namespace so they could probably only disable jobs that they have access too.

In MultiKueue changing the label is up to the Kueue controller. Once the job is running it cannot be flipped, because not mutable then (as other labels).

> Last thing, what is the expected behavior if a label is added on an existing job? Will we stop syncing the running job?

it is not possible to add the label to a running job, because `AllowMutableSchedulingDirectives=false` in this case.

Reply:
So I guess I was confused on where the label is located. If its on the label of a job object (ie metadata) then it is anyone is free to patch or modify anytime. If the label is added on the PodTemplate metadata then that would be covered by `AllowMutableSchedulingDirectives`. That is immutable once a job is running.

Reply:
Oh, I was actually confused :). Indeed `AllowMutableSchedulingDirectives` applies only to the template currently, while the job will be placed on the top level metadata. Thank you for pointing this out!

In that case I will need to rethink. We either make it immutable, or extend the semantics of `AllowMutableSchedulingDirectives` to cover the particular label in metadata. WDYT?

Reply:
I think the semantic of `AllowMutableSchedulingDirectives` is actually what we want. Rephrased the "Mutability" section to make it more clear. PTAL.

Reply:
After a second thought I think I prefer to keep it simple and make the label immutable, at least for the first iteration. Added a graduation criteria to re-evaluate the decision.

Reply:
Yea I think that is fine. So the crux of the change is to define a label, have controllers that create Jobs can populate it, and the Job controller will skip sync.

And we won't allow any changes to this label once the job is created.

Reply:
I would consider adding this to Risks and Mitigation section (as without immutable) this could render the jobs useless.

Reply:
I'm not sure it should go under Risks and Mitigations since this is a hypothetical risk if the design was different. I put a discussion under the "Notes / Can the label be mutable?" section. PTAL.

Reply:
Thanks. Your updates look good to me! Feel free to resolve.